### ISSUE TITLE
Run tests against pip installed version, not from repo

### ISSUE THREAD

**Issue by pganssle** [dateutil/dateutil#506]
Currently the tests are just run from the repo on travis and appveyor, I guess. We should change the CI script so that the tests are run against a pip installed version.

It may be sufficient to just do `mkdir build_test && cd build_test` then `pytest ..` so that `dateutil` is not in `.`

If necessary, we can move `test` out of `dateutil/` into a sibling directory of `dateutil`.

**Comment by pganssle** [comment]
Here is how `pandas` does it, according to https://github.com/pandas-dev/pandas/issues/18141#issuecomment-342954600:  https://github.com/pandas-dev/pandas/blob/master/ci/install_travis.sh#L158

**Comment by pganssle** [comment]
[A relevant blog post](https://enotuniq.org/sourcing_your_source_without_src.html)

**Comment by wagner-certat** [comment]
> Currently the tests are just run from the repo on travis and appveyor, I guess. We should change the CI script so that the tests are run against a pip installed version.

Aren't they the same?

**Comment by pganssle** [comment]
In theory they are the same, but you want to work against the pip installed version because you want to make sure that the `setup.py` script is working as expected.

The bug that's happening now is that when you try to do `from dateutil import parser`, it fails in the pip installed version, because `setup.py` isn't installing the `dateutil.parser` submodule (which was not listed in the `packages` option). However, the tests are all passing, because in the *repo*, the `parser` folder exists, and the test script is picking it up from there.

If we renamed `dateutil` to `src`, we wouldn't have this problem because `from dateutil import parser` wouldn't find anything in the CWD, but I am worried that if we do that it will break something in an unexpected way.

**Comment by wagner-certat** [comment]
> because you want to make sure that the `setup.py` script is working as expected.

Or use `setup.py test`, then you can check both things in one run :D

**Comment by pganssle** [comment]
@wagner-certat I don't think so. The problem is this:

```
[dateutil] $ virtualenv venv
Using base prefix '/usr'
New python executable in /home/c0zj/Documents/Programming/Python/dateutil/dateutil/venv/bin/python
Installing setuptools, pip, wheel...done.

[dateutil] $ source venv/bin/activate

(venv) [dateutil] $ pip install .
Processing ~/Programming/Python/dateutil
Collecting six>=1.5 (from python-dateutil==2.6.1)
  Using cached six-1.11.0-py2.py3-none-any.whl
Installing collected packages: six, python-dateutil
  Running setup.py install for python-dateutil ... done
Successfully installed python-dateutil-2.6.1 six-1.11.0

(venv) [dateutil] $ python -c 'from dateutil import parser; print("Success")'
Success

(venv) [dateutil] $ cd venv

(venv) [dateutil/venv] $ python -c 'from dateutil import parser; print("Success")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: cannot import name 'parser'
```

In the first `python` call, `PYTHONPATH` finds `dateutil` as a module in my current working directory and finds `parser` in there (because `parser` is in the repo), and imports from there. In the *second* call, it doesn't see `dateutil` in my current working directory so it looks in `site-packages`, finds the installed version of `dateutil`, but doesn't find `parser` in there, because in `setup.py` we didn't include it in the `packages` line. Obviously I can fix this by just adding `parser` to the packages line, but this has happened before (with the `tz` module), and I didn't set up CI to test for it then, so I think we should set up CI to test for it *now*, so it won't happen a third time.

To make it obvious, you can compare the repo version of `dateutil`:

```
(venv) [dateutil] $ tree dateutil
dateutil
├── _common.py
├── easter.py
├── __init__.py
├── parser
│   ├── __init__.py
│   ├── _parser.py
│   └── __pycache__
│       ├── __init__.cpython-36.pyc
│       └── _parser.cpython-36.pyc
├── __pycache__
│   ├── _common.cpython-36.pyc
│   ├── __init__.cpython-36.pyc
│   ├── relativedelta.cpython-36.pyc
│   └── _version.cpython-36.pyc
├── relativedelta.py
├── rrule.py
├── test
│   ├── _common.py
│   ├── __init__.py
│   ├── test_easter.py
│   ├── test_imports.py
│   ├── test_import_star.py
│   ├── test_internals.py
│   ├── test_parser.py
│   ├── test_parser_pytest.py
│   ├── test_relativedelta.py
│   ├── test_rrule.py
│   ├── test_tz.py
│   └── test_utils.py
├── tz
│   ├── _common.py
│   ├── __init__.py
│   ├── __pycache__
│   │   ├── _common.cpython-36.pyc
│   │   ├── __init__.cpython-36.pyc
│   │   ├── tz.cpython-36.pyc
│   │   └── win.cpython-36.pyc
│   ├── tz.py
│   └── win.py
├── tzwin.py
├── utils.py
├── _version.py
└── zoneinfo
    ├── dateutil-zoneinfo.tar.gz
    ├── __init__.py
    └── rebuild.py
```

To the installed version:

```
(venv) [dateutil] $ tree venv/lib/python3.6/site-packages/dateutil/
venv/lib/python3.6/site-packages/dateutil/
├── _common.py
├── easter.py
├── __init__.py
├── __pycache__
│   ├── _common.cpython-36.pyc
│   ├── easter.cpython-36.pyc
│   ├── __init__.cpython-36.pyc
│   ├── relativedelta.cpython-36.pyc
│   ├── rrule.cpython-36.pyc
│   ├── tzwin.cpython-36.pyc
│   ├── utils.cpython-36.pyc
│   └── _version.cpython-36.pyc
├── relativedelta.py
├── rrule.py
├── tz
│   ├── _common.py
│   ├── __init__.py
│   ├── __pycache__
│   │   ├── _common.cpython-36.pyc
│   │   ├── __init__.cpython-36.pyc
│   │   ├── tz.cpython-36.pyc
│   │   └── win.cpython-36.pyc
│   ├── tz.py
│   └── win.py
├── tzwin.py
├── utils.py
├── _version.py
└── zoneinfo
    ├── dateutil-zoneinfo.tar.gz
    ├── __init__.py
    ├── __pycache__
    │   ├── __init__.cpython-36.pyc
    │   └── rebuild.cpython-36.pyc
    └── rebuild.py
```

You can see the installed version is missing the `test` module (deliberately) and the `parser` module (accidentally).

**Comment by pganssle** [comment]
Fixed in #509.

This is *not* fixed in `tox.ini`, but mainly because it's a real pain and CI will catch these issues anyway.
